lib/postgrex/replication.ex

```diff
@@ -694,140 +439,45 @@ defmodule Postgrex.Replication do
-    begin_statement = command(:begin_copy, opts)
-    create_slot_statement = command(:create_slot, opts)
-    copy_statement = command(:copy_table, opts)
-
-    with {:ok, _, protocol} <- Protocol.handle_simple(begin_statement, s.protocol),
-         {:ok, slot, protocol} <- Protocol.handle_simple(create_slot_statement, protocol),
-         {:ok, columns, protocol} <- table_columns(table_name, protocol),
-         {:ok, protocol} <- Protocol.handle_copy_table(copy_statement, protocol),
-         {:ok, protocol} <- Protocol.checkin(protocol) do
-      s = %{s | protocol: protocol, copy_opts: Keyword.put(opts, :columns, columns)}
-      {:reply, {:ok, slot}, s}
-    else
-      {:error, reason, protocol} ->
-        {:reply, {:error, reason}, %{s | protocol: protocol}}
-
-      {:disconnect, reason, protocol} ->
-        reconnect_or_stop(:disconnect, reason, protocol, s)
-    end
-  end
-
-  @doc false
-  def handle_call({name, opts}, _from, %{repl_opts: nil, copy_opts: nil} = s)
-      when name in @public_commands do
-    statement = command(name, opts)
-
-    case Protocol.handle_simple(statement, s.protocol) do
-      {:ok, %Postgrex.Result{} = result, protocol} ->
-        {:reply, {:ok, result}, %{s | protocol: protocol}}
-
-      {:error, reason, protocol} ->
-        {:reply, {:error, reason}, %{s | protocol: protocol}}
-
-      {:disconnect, reason, protocol} ->
-        reconnect_or_stop(:disconnect, reason, protocol, s)
-    end
-  end
-
-  @doc false
-  def handle_call({name, _opts}, _from, s) when name in @public_commands,
-    do: {:reply, {:error, :stream_in_progress}, s}
-
-  @doc false
   def handle_call(msg, from, %{state: {mod, mod_state}} = s) do
-    handle(mod, :handle_call, [msg, from, mod_state], s)
+    handle(mod, :handle_call, [msg, from, mod_state], from, s)
   end
 
   @doc false
-  def handle_info(msg, %{protocol: protocol} = s) do
-    max_copies = max_copy_messages(s)
-
-    case Protocol.handle_copy_recv(msg, max_copies, protocol) do
+  def handle_info(msg, %{protocol: protocol, streaming: streaming} = s) do
+    case Protocol.handle_copy_recv(msg, streaming, protocol) do
       {:ok, copies, protocol} ->
-        copy_recv_dispatch(copies, %{s | protocol: protocol})
+        handle_data(copies, %{s | protocol: protocol})
 
       :unknown ->
         %{state: {mod, mod_state}} = s
-
-        if function_exported?(mod, :handle_info, 2) do
-          handle(mod, :handle_info, [msg, mod_state], s)
-        else
-          {:noreply, s}
-        end
+        maybe_handle(mod, :handle_info, [msg, mod_state], s)
 
       {error, reason, protocol} ->
         reconnect_or_stop(error, reason, protocol, s)
     end
   end
 
-  defp copy_recv_dispatch(copies, %{copy_opts: nil} = s), do: handle_data(copies, s)
-  defp copy_recv_dispatch(copies, s), do: handle_copy(copies, s)
-
   defp handle_data([], s), do: {:noreply, s}
 
   defp handle_data([copy | copies], %{state: {mod, mod_state}} = s) do
```

Comment:
In the `Postgrex.Protocol` module, the `handle_copy_recv` function is signalling the end of the regular copying:

```Elixir
{:ok, msg_ready(status: postgres), buffer} ->
  s = %{s | postgres: postgres, buffer: buffer}
  {:ok, Enum.reverse([:copy_done | copies]), s}
```

This could be used to reset the state/exit streaming after copying is done. Just need to document + add it to the type spec for `handle_data`.

I was reading more and it seems like having this sort of signal is also crucial when starting replication on a timeline that's not the latest one. [From the Postgres docs](https://www.postgresql.org/docs/current/protocol-replication.html):


> After streaming all the WAL on a timeline that is not the latest one, the server will end streaming by exiting the COPY mode. 
>
> When the client acknowledges this by also exiting COPY mode, the server sends a result set with one row and two columns, indicating the next timeline in this server's history. 
>
> The first column is the next timeline's ID (type int8), and the second column is the WAL location where the switch happened (type text). 
>
> Finally, the server sends two CommandComplete messages (one that ends the CopyData and the other ends the START_REPLICATION itself), and is ready to accept a new command.